external/concealer/src/autoware.cpp

```diff
@@ -13,37 +13,18 @@
 // limitations under the License.
 
 #include <concealer/autoware.hpp>
-#include <cstdlib>
-#include <exception>
-#include <scenario_simulator_exception/exception.hpp>
 
 namespace concealer
 {
-void Autoware::checkAutowareProcess()
+Autoware::Autoware()
+: rclcpp::Node("concealer", "simulation", rclcpp::NodeOptions().use_global_arguments(false))
 {
-  if (process_id != 0) {
-    int wstatus = 0;
-    int ret = waitpid(process_id, &wstatus, WNOHANG);
-    if (ret == 0) {
-      return;
-    } else if (ret < 0) {
-      if (errno == ECHILD) {
-        is_autoware_exited = true;
-        throw common::AutowareError("Autoware process is already terminated");
-      } else {
-        AUTOWARE_SYSTEM_ERROR("waitpid");
-        std::exit(EXIT_FAILURE);
-      }
-    }
+}
 
-    if (WIFEXITED(wstatus)) {
-      is_autoware_exited = true;
-      throw common::AutowareError(
-        "Autoware process is unintentionally exited. exit code: ", WEXITSTATUS(wstatus));
-    } else if (WIFSIGNALED(wstatus)) {
-      is_autoware_exited = true;
-      throw common::AutowareError("Autoware process is killed. signal is ", WTERMSIG(wstatus));
-    }
+auto Autoware::spinSome() -> void
+{
+  if (rclcpp::ok()) {
+    rclcpp::spin_some(get_node_base_interface());
```

Comment:
spin_once will execute all callbacks from a Subscription Queue. If the data did not reach the subscription queue until spin_once is called, it would not be available for this frame of traffic_simlulator execution but will be available for the next.  

What do you mean by "we will receive all the necessary data from Autoware"? The traffic simulator could wait until each of the data subscribed by the concealer is updated at least once between frames, but it would block the execution longer than necessary. traffic_simulator works with its specific frequency, and Autoware publishes data with its frequencies. Some data will be updated between EgoEntity::onUpdate calls, and some will not. If they were not updated, the old value is used to proceed with the simulation.

Please let me know if I misunderstood something.

Reply:
Thanks!
I understand the specification of spin_once.